# Calculating Across Rows

How do we calculate row-wise in Pandas like we do in Excel?

<img src="https://sandeepmj.github.io/image-host/oil-v-gold.png">


We need a couple of new concepts to do this calculation:

1. Capture a number from a cell and hold in variable.
2. We `apply()`calculations across columns, row-wise, using using a `lambda expression`.


This is step one in your ability to accomplish tasks that are highly specialized to your beats.



In [103]:
## import library
import pandas as pd

In [104]:
## pull in data
df = pd.read_csv("https://raw.githubusercontent.com/sandeepmj/datasets/refs/heads/main/oil-v-gold-pandas.csv")
df.head()

,Date,PX_GOLD,PX_OIL
0,1/31/08,922.7,91.75
1,2/29/08,975.0,101.84
2,3/31/08,916.2,101.58
3,4/30/08,865.1,113.46
4,5/30/08,887.3,127.35


What are the different values for row **TWO** in this formula for `PX_GOLD`:

$$\text{Pct Chg at time } t = \frac{\text{Value at time } t - \text{Value at start time}}{\text{Value at start time}} \times 100$$

Use `.iat` to store a value from a specific cell into a variable:

```dataframe["column_name"].iat[row_index_number]```

In [105]:
## hold gold
df["PX_GOLD"].iat[0]

np.float64(922.7)

Also works:

`dataframe.iat[col_index_number, row_index_number]`

In [106]:
## hold gold col index, row index
gold_start = df.iat[0,1]
gold_start

np.float64(922.7)

Place the value for the starting oil price in a variable called `oil_baseline`

In [107]:
## oil_baseline
oil_start = df.iat[0,2]
oil_start

np.float64(91.75)

# `lambda expressions`

We're going to just write one first and then break it down:

In [108]:
## create a new column that holds the zero index for gold
df["gold_zero_index"] = df["PX_GOLD"].apply(lambda x: round( ( (x - gold_start) / gold_start ) * 100,1 ) )
df.head()

,Date,PX_GOLD,PX_OIL,gold_zero_index
0,1/31/08,922.7,91.75,0.0
1,2/29/08,975.0,101.84,5.7
2,3/31/08,916.2,101.58,-0.7
3,4/30/08,865.1,113.46,-6.2
4,5/30/08,887.3,127.35,-3.8


## What  `x` stand for:

- In Pandas, we `apply()` lambda expressions to columns.
-  `.apply()`  works element-wise on each value in the `PX_GOLD` column, one at a time.
- `x` is a single value representing a cell in the column `PX_GOLD`
- The ```lambda x:``` processes each row of a dataframe, one row at a time.
-  The `lambda expression` takes each individual number `x` from `PX_GOLD` and runs the indexing formula on it.

Think of it as: "call each value in `PX_GOLD` as `x`, and do this math to using `x` in the formula." 

## create a new column that holds the zero index for oil



In [109]:
## zero index oil

df["oil_zero_index"] = df["PX_OIL"].apply(lambda x: round( ( (x - oil_start) / oil_start) * 100,1 ) )
df.head()

,Date,PX_GOLD,PX_OIL,gold_zero_index,oil_zero_index
0,1/31/08,922.7,91.75,0.0,0.0
1,2/29/08,975.0,101.84,5.7,11.0
2,3/31/08,916.2,101.58,-0.7,10.7
3,4/30/08,865.1,113.46,-6.2,23.7
4,5/30/08,887.3,127.35,-3.8,38.8


In [110]:
## call df
df

,Date,PX_GOLD,PX_OIL,gold_zero_index,oil_zero_index
0,1/31/08,922.7,91.75,0.0,0.0
1,2/29/08,975.0,101.84,5.7,11.0
2,3/31/08,916.2,101.58,-0.7,10.7
3,4/30/08,865.1,113.46,-6.2,23.7
4,5/30/08,887.3,127.35,-3.8,38.8
...,...,...,...,...,...
116,9/29/17,1281.5,51.67,38.9,-43.7
117,10/31/17,1270.5,54.38,37.7,-40.7
118,11/30/17,1273.2,57.40,38.0,-37.4
119,12/29/17,1309.3,60.42,41.9,-34.1


### 100-indexing

Create cells below to add columns for both oil and gold for 100-indexing.

Reminder, the formula for 100-indexing:

$$\text{Index} = \left(\frac{\text{value at time } t}{\text{value at start}}\right) \times 100$$

In [111]:
## gold 100

df["gold_100_index"] = df["PX_GOLD"].apply(lambda x: round( ( x / gold_start ) * 100,1 ) )
df.head()

,Date,PX_GOLD,PX_OIL,gold_zero_index,oil_zero_index,gold_100_index
0,1/31/08,922.7,91.75,0.0,0.0,100.0
1,2/29/08,975.0,101.84,5.7,11.0,105.7
2,3/31/08,916.2,101.58,-0.7,10.7,99.3
3,4/30/08,865.1,113.46,-6.2,23.7,93.8
4,5/30/08,887.3,127.35,-3.8,38.8,96.2


In [112]:
## oil 100

df["oil_100_index"] = df["PX_OIL"].apply(lambda x: round( ( x / oil_start ) * 100,1 ) )
df.head()

,Date,PX_GOLD,PX_OIL,gold_zero_index,oil_zero_index,gold_100_index,oil_100_index
0,1/31/08,922.7,91.75,0.0,0.0,100.0,100.0
1,2/29/08,975.0,101.84,5.7,11.0,105.7,111.0
2,3/31/08,916.2,101.58,-0.7,10.7,99.3,110.7
3,4/30/08,865.1,113.46,-6.2,23.7,93.8,123.7
4,5/30/08,887.3,127.35,-3.8,38.8,96.2,138.8


# Homework

<a href="https://raw.githubusercontent.com/sandeepmj/datasets/refs/heads/main/demo-data-solutions/cablebill-v-income.csv">This dataset</a> shows the price of a standard cable subscription and the U.S. median income annually between 1995-2023.

Has median income kept pace with the price of a cable subscription?

Round all numbers to 1 decimal place



In [113]:
## read file
dfHW = pd.read_csv("https://raw.githubusercontent.com/sandeepmj/datasets/refs/heads/main/demo-data-solutions/cablebill-v-income.csv")
dfHW.head()

,year,avg_cable_bill,median_household_income
0,1995,22.35,35249
1,1996,23.28,36689
2,1997,26.31,38229
3,1998,27.88,40148
4,1999,28.94,41994


In [114]:
## get info
dfHW.info()

<class 'pandas.DataFrame'>
RangeIndex: 29 entries, 0 to 28
Data columns (total 3 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   year                     29 non-null     int64  
 1   avg_cable_bill           29 non-null     float64
 2   median_household_income  29 non-null     int64  
dtypes: float64(1), int64(2)
memory usage: 828.0 bytes


### Index cable billl

In [115]:
## cable_baseline

cable_start = dfHW["avg_cable_bill"].iat[0]
cable_start

np.float64(22.35)

In [116]:
## cable bill index

dfHW["cable_zero_index"] = dfHW["avg_cable_bill"].apply(lambda x: round( ((x - cable_start) /cable_start) * 100,1 ) )
dfHW.head()

,year,avg_cable_bill,median_household_income,cable_zero_index
0,1995,22.35,35249,0.0
1,1996,23.28,36689,4.2
2,1997,26.31,38229,17.7
3,1998,27.88,40148,24.7
4,1999,28.94,41994,29.5


### Index median income

In [117]:
## income_baseline

income_start = dfHW["median_household_income"].iat[0]
income_start

np.int64(35249)

In [121]:
## median income index

dfHW["income_zero_index"] = dfHW["median_household_income"].apply(lambda x: round( ( (x - income_start) / income_start) * 100,1 ) )
dfHW

,year,avg_cable_bill,median_household_income,cable_zero_index,income_zero_index
0,1995,22.35,35249,0.0,0.0
1,1996,23.28,36689,4.2,4.1
2,1997,26.31,38229,17.7,8.5
3,1998,27.88,40148,24.7,13.9
4,1999,28.94,41994,29.5,19.1
5,2000,31.22,42349,39.7,20.1
6,2001,33.75,42710,51.0,21.2
7,2002,36.47,43057,63.2,22.2
8,2003,38.95,43564,74.3,23.6
9,2004,41.04,44684,83.6,26.8


In [119]:
## analysis

Between 1995 and 2023, cable bill prices in the United States rose by more than 1,200%. In that same time, the median household income rose just by about 113% nationally.